In [1]:
from numpy import* 
from pandas import*
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
credit_card_data= read_csv(r"C:\Users\Asus\Downloads\creditcard.csv")

In [3]:
df=DataFrame(credit_card_data)

In [4]:
df.head(5)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [6]:
df['Class'].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [7]:
# 0 -----> Normal Transaction
# 1 -----> Fraud Transaction

In [8]:
df_Normal=df[df.Class==0]
df_Fraud=df[df.Class==1]

In [9]:
df_Fraud.head()

Time        V1        V2        V3        V4        V5        V6  \
541    406.0 -2.312227  1.951992 -1.609851  3.997906 -0.522188 -1.426545   
623    472.0 -3.043541 -3.157307  1.088463  2.288644  1.359805 -1.064823   
4920  4462.0 -2.303350  1.759247 -0.359745  2.330243 -0.821628 -0.075788   
6108  6986.0 -4.397974  1.358367 -2.592844  2.679787 -1.128131 -1.706536   
6329  7519.0  1.234235  3.019740 -4.304597  4.732795  3.624201 -1.357746   

            V7        V8        V9  ...       V21       V22       V23  \
541  -2.537387  1.391657 -2.770089  ...  0.517232 -0.035049 -0.465211   
623   0.325574 -0.067794 -0.270953  ...  0.661696  0.435477  1.375966   
4920  0.562320 -0.399147 -0.238253  ... -0.294166 -0.932391  0.172726   
6108 -3.496197 -0.248778 -0.247768  ...  0.573574  0.176968 -0.436207   
6329  1.713445 -0.496358 -1.282858  ... -0.379068 -0.704181 -0.656805   

           V24       V25       V26       V27       V28  Amount  Class  
541   0.320198  0.044519  0.177840  0.261145 -0.143276    0.00      1  
623  -0.293803  0.279798 -0.145362 -0.252773  0.035764  529.00      1  
4920 -0.087330 -0.156114 -0.542628  0.039566 -0.153029  239.93      1  
6108 -0.053502  0.252405 -0.657488 -0.827136  0.849573   59.00      1  
6329 -1.632653  1.488901  0.566797 -0.010016  0.146793    1.00      1  

[5 rows x 31 columns]

In [10]:
df_Fraud.shape

(492, 31)

In [11]:
df_Normal.shape

(284315, 31)

In [12]:
df_Fraud["Amount"].describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [13]:
df_Normal["Amount"].describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

#Compare these two data 

In [14]:
df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

Buid a sample dataset containing similar distribution of normal transaction and the fraudulent transaction 

Number of Fraudulent Transactions====> 492
so we are taking 492 normal transaction for balance data

In [15]:
legit=df_Normal.sample(n=492) #it randomly sample the data

In [16]:
legit

Time        V1        V2        V3        V4        V5        V6  \
43477    41524.0  1.026188 -0.347634  1.338286  0.475906 -0.824275  0.843391   
153798  100127.0  1.855546 -1.382218 -0.613522 -1.282491 -1.294309 -0.955857   
27469    34561.0  0.241152 -1.463549 -0.896442  1.353269  0.068698  0.764085   
239339  150058.0 -1.393951 -0.139982 -0.367128 -2.247680  4.423103  2.707322   
47757    43348.0 -0.019284  0.651875  0.356541  0.193422  0.448395 -0.458365   
...          ...       ...       ...       ...       ...       ...       ...   
207588  136744.0 -0.868208 -0.158921  1.624708 -3.016360 -1.079055 -0.749061   
264871  161634.0 -2.707969 -2.681048  1.616168 -0.705674  3.171417 -1.981312   
11034    19033.0 -0.338866  1.155634 -0.094423  1.133782 -0.182395 -0.709607   
68637    53076.0  1.187824  0.205501  0.148704  0.456981 -0.009393 -0.099744   
218157  141190.0 -1.219627  1.706352 -1.254104 -0.461239  1.203430 -0.408042   

              V7        V8        V9  ...       V21       V22       V23  \
43477  -1.037225  0.496998  0.481489  ...  0.341903  1.030581 -0.002124   
153798 -0.566674 -0.394392  1.904999  ... -0.816558 -1.700378  0.383504   
27469   0.743206  0.031440 -0.077875  ...  0.270881 -0.258485 -0.719227   
239339  0.393657  0.769951 -0.426803  ... -0.331131 -1.248758 -0.168596   
47757   0.344884 -0.167355 -0.014469  ...  0.050462  0.119408  0.046770   
...          ...       ...       ...  ...       ...       ...       ...   
207588 -0.136288  0.008061 -1.644969  ... -0.244935 -0.171571 -0.334373   
264871 -1.633829  0.450659  0.635774  ...  0.540386  0.736696  0.202431   
11034   0.675801  0.268873  0.485985  ...  0.113670  0.416149  0.286180   
68637  -0.127832  0.127690 -0.140885  ... -0.233398 -0.701689  0.108798   
218157  1.124494  0.646935 -1.467798  ...  0.345624  0.563954 -0.431149   

             V24       V25       V26       V27       V28  Amount  Class  
43477  -0.240583  0.022928  0.511473  0.047038  0.018641   25.79      0  
153798 -0.029763 -0.669878 -0.254370 -0.036927 -0.017406  166.56      0  
27469  -1.126767  0.677889 -0.228387 -0.081680  0.076330  518.08      0  
239339  0.588453  0.698833  0.132627 -0.248079 -0.157384    9.99      0  
47757  -0.587260 -1.163727  0.293514 -0.108724  0.188624    4.69      0  
...          ...       ...       ...       ...       ...     ...    ...  
207588 -0.012245  0.675596 -0.112102  0.387880  0.171611   28.40      0  
264871 -0.484285  0.199849 -0.412862  0.029394  0.196356    9.99      0  
11034   0.238732 -0.747224 -0.467533  0.245625  0.155424   89.99      0  
68637  -0.360504  0.140867  0.128242 -0.014278  0.013410    4.49      0  
218157  0.075943  0.690991 -0.547014 -0.306203 -0.033593   36.00      0  

[492 rows x 31 columns]

In [17]:
new_df= concat(([legit,df_Fraud]), axis=0)

In [18]:
new_df

Time        V1        V2        V3        V4        V5        V6  \
43477    41524.0  1.026188 -0.347634  1.338286  0.475906 -0.824275  0.843391   
153798  100127.0  1.855546 -1.382218 -0.613522 -1.282491 -1.294309 -0.955857   
27469    34561.0  0.241152 -1.463549 -0.896442  1.353269  0.068698  0.764085   
239339  150058.0 -1.393951 -0.139982 -0.367128 -2.247680  4.423103  2.707322   
47757    43348.0 -0.019284  0.651875  0.356541  0.193422  0.448395 -0.458365   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
43477  -1.037225  0.496998  0.481489  ...  0.341903  1.030581 -0.002124   
153798 -0.566674 -0.394392  1.904999  ... -0.816558 -1.700378  0.383504   
27469   0.743206  0.031440 -0.077875  ...  0.270881 -0.258485 -0.719227   
239339  0.393657  0.769951 -0.426803  ... -0.331131 -1.248758 -0.168596   
47757   0.344884 -0.167355 -0.014469  ...  0.050462  0.119408  0.046770   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
43477  -0.240583  0.022928  0.511473  0.047038  0.018641   25.79      0  
153798 -0.029763 -0.669878 -0.254370 -0.036927 -0.017406  166.56      0  
27469  -1.126767  0.677889 -0.228387 -0.081680  0.076330  518.08      0  
239339  0.588453  0.698833  0.132627 -0.248079 -0.157384    9.99      0  
47757  -0.587260 -1.163727  0.293514 -0.108724  0.188624    4.69      0  
...          ...       ...       ...       ...       ...     ...    ...  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[984 rows x 31 columns]

In [19]:
new_df['Class'].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [20]:
new_df.groupby('Class').mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94734.180894 -0.040131  0.094429 -0.090450  0.121561  0.009571   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0     -0.025403 -0.006690  0.009147 -0.025964  ... -0.023614  0.019968   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0      0.042720 -0.082070  0.007098  0.019447 -0.001619  0.014152  0.013149   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       82.570407  
1      122.211321  

[2 rows x 30 columns]

In [21]:
X= new_df.drop(columns='Class',axis=1)
Y= new_df["Class"]

In [22]:
X

Time        V1        V2        V3        V4        V5        V6  \
43477    41524.0  1.026188 -0.347634  1.338286  0.475906 -0.824275  0.843391   
153798  100127.0  1.855546 -1.382218 -0.613522 -1.282491 -1.294309 -0.955857   
27469    34561.0  0.241152 -1.463549 -0.896442  1.353269  0.068698  0.764085   
239339  150058.0 -1.393951 -0.139982 -0.367128 -2.247680  4.423103  2.707322   
47757    43348.0 -0.019284  0.651875  0.356541  0.193422  0.448395 -0.458365   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V20       V21       V22  \
43477  -1.037225  0.496998  0.481489  ... -0.090874  0.341903  1.030581   
153798 -0.566674 -0.394392  1.904999  ... -0.252978 -0.816558 -1.700378   
27469   0.743206  0.031440 -0.077875  ...  0.850048  0.270881 -0.258485   
239339  0.393657  0.769951 -0.426803  ...  0.200988 -0.331131 -1.248758   
47757   0.344884 -0.167355 -0.014469  ... -0.070163  0.050462  0.119408   
...          ...       ...       ...  ...       ...       ...       ...   
279863 -0.882850  0.697211 -2.064945  ...  1.252967  0.778584 -0.319189   
280143 -1.413170  0.248525 -1.127396  ...  0.226138  0.370612  0.028234   
280149 -2.234739  1.210158 -0.652250  ...  0.247968  0.751826  0.834108   
281144 -2.208002  1.058733 -1.632333  ...  0.306271  0.583276 -0.269209   
281674  0.223050 -0.068384  0.577829  ... -0.017652 -0.164350 -0.295135   

             V23       V24       V25       V26       V27       V28  Amount  
43477  -0.002124 -0.240583  0.022928  0.511473  0.047038  0.018641   25.79  
153798  0.383504 -0.029763 -0.669878 -0.254370 -0.036927 -0.017406  166.56  
27469  -0.719227 -1.126767  0.677889 -0.228387 -0.081680  0.076330  518.08  
239339 -0.168596  0.588453  0.698833  0.132627 -0.248079 -0.157384    9.99  
47757   0.046770 -0.587260 -1.163727  0.293514 -0.108724  0.188624    4.69  
...          ...       ...       ...       ...       ...       ...     ...  
279863  0.639419 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00  
280143 -0.145640 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76  
280149  0.190944  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89  
281144 -0.456108 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00  
281674 -0.072173 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53  

[984 rows x 30 columns]

split it into Training Data and Testing Data

In [23]:
X_train, X_test, Y_train, Y_test= train_test_split(X,Y, test_size=0.2, stratify=Y, random_state=2)

In [24]:
X_test

Time         V1         V2         V3         V4         V5  \
247995  153761.0   1.146259   1.403458  -4.159148   2.660107  -0.323217   
264871  161634.0  -2.707969  -2.681048   1.616168  -0.705674   3.171417   
230902  146528.0   2.013082  -1.187894  -0.277178  -0.661384  -0.966359   
200176  133320.0   2.286248  -0.567130  -2.615480  -0.933238   0.254480   
15451    26833.0 -20.532751  12.373989 -23.009003   6.144821 -15.587296   
...          ...        ...        ...        ...        ...        ...   
108258   70828.0   0.196707   1.189757   0.704882   2.891388   0.045555   
12108    21046.0 -16.917468   9.669900 -23.736443  11.824990  -9.830548   
99506    67150.0  -1.824295   0.403327  -1.994122   2.756558  -3.139064   
74496    55614.0  -7.347955   2.397041  -7.572356   5.177819  -2.854838   
116539   74321.0   0.306120   1.775322  -1.479935   0.801335   1.828580   

              V6         V7         V8        V9  ...       V20       V21  \
247995 -1.836071  -1.623740   0.259562 -1.132044  ...  0.284831  0.564450   
264871 -1.981312  -1.633829   0.450659  0.635774  ...  0.567899  0.540386   
230902  0.668924  -1.382116   0.248352  0.011341  ... -0.395746 -0.283112   
200176 -1.539013   0.365753  -0.545431 -0.843306  ... -0.113341  0.459180   
15451  -4.384491 -15.939003  13.696416 -3.948455  ...  1.592754  1.754608   
...          ...        ...        ...       ...  ...       ...       ...   
108258  1.245730  -1.198714  -2.421616 -1.232089  ...  0.646616 -1.328132   
12108  -2.514829 -17.290657   1.820408 -6.264903  ...  0.996745 -2.336111   
99506   0.408185  -1.209045   1.095634 -1.447225  ...  1.503688  0.838760   
74496  -1.795239  -8.783235   0.437157 -3.740598  ...  0.073164 -0.175273   
116539 -0.459795   0.738484  -0.998311 -1.083583  ... -0.015206  0.546883   

             V22       V23       V24       V25       V26       V27       V28  \
247995  0.445744 -0.141136 -0.265517  0.362260 -0.416062  0.507370  0.243744   
264871  0.736696  0.202431 -0.484285  0.199849 -0.412862  0.029394  0.196356   
230902 -0.334598  0.299665  0.091612 -0.632869  0.467006 -0.000032 -0.033601   
200176  1.281242 -0.192389  0.795952  0.736054  0.270561 -0.112310 -0.087469   
15451  -1.466115 -0.856779  0.125777  1.402587 -0.223755  1.574249  0.469201   
...          ...       ...       ...       ...       ...       ...       ...   
108258  0.189311 -0.005524 -0.814708  0.400924  0.286281  0.135215  0.257315   
12108   0.972755  1.241866 -1.051086  0.038009  0.672317  2.108471 -1.421243   
99506   0.341727  0.947506 -0.145493  0.049326  0.831065  0.332421  0.252713   
74496   0.543325 -0.547955 -0.503722 -0.310933 -0.163986  1.197895  0.378187   
116539 -0.791544 -0.535405 -1.324555  1.314524 -0.206324  0.200530  0.280182   

        Amount  
247995   51.37  
264871    9.99  
230902   53.10  
200176   24.00  
15451    99.99  
...        ...  
108258    0.76  
12108     1.00  
99506   489.71  
74496     0.83  
116539    0.74  

[197 rows x 30 columns]

In [25]:
model = LogisticRegression()

In [26]:
model.fit(X_train,Y_train)

LogisticRegression()

In [27]:
X_train_pre= model.predict(X_train)
training_data_accuracy= accuracy_score(X_train_pre,Y_train)

In [28]:
training_data_accuracy

0.9428208386277002

In [29]:
X_test_pre= model.predict(X_test)

In [30]:
testing_data_accuracy= accuracy_score(X_test_pre, Y_test)
testing_data_accuracy

0.934010152284264

Model prediction is good as it shows 93.4% accuracy